### PR TITLE
Add support for specifying execution cluster labels in pyflyte

### PR DESCRIPTION
**Pull request by va6996**
## Tracking issue

_NA_

## Why are the changes needed?
Today we support the execution cluster label for specifying the execution cluster on an execution level. However there is no way to do that using the python sdk/pyflyte. This change adds support for the above mentioned.

## What changes were proposed in this pull request?
We add another flag to pyflyte to allow users to specify the cluster label. Additionally when launching a wf, this info is passed to flyteadmin via the sdk.

## How was this patch tested?
This patch was tested on a local instance of flyte. Test cases were modified to validate the changes.

### Setup process
`pyflyte run -ecl my-label mywf.py wf`

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

**Comment by Tom-Newton** [comment]
Thanks for making the PR. I have not looked at it in detail but I think we need to upgrade the required version of `flyteidl`. I think at least version `v1.11.1-b1` is needed here https://github.com/flyteorg/flytekit/blob/76fb7c344162b7fae141c40c2fc8d18a71091fc2/pyproject.toml#L23

File: flytekit/clis/sdk_in_container/run.py [hunk 1]
```diff
@@ -239,7 +239,7 @@ class RunLevelParams(PyFlyteParams):
     )
     limit: int = make_click_option_field(
         click.Option(
-            param_decls=["--limit", "limit"],
+            param_decls=["--limit"],
```
Comment (fg91) [inline flytekit/clis/sdk_in_container/run.py:242]: Nit: bit of scope creep ;)

Still valid of course.

File: flytekit/clis/sdk_in_container/run.py [hunk 2]
```diff
@@ -249,13 +249,23 @@ class RunLevelParams(PyFlyteParams):
     )
     cluster_pool: str = make_click_option_field(
         click.Option(
-            param_decls=["--cluster-pool", "cluster_pool"],
+            param_decls=["--cluster-pool"],
```
Comment (fg91) [inline flytekit/clis/sdk_in_container/run.py:252]: All the other options support an underscore version and technically this is backwards breaking so I'd vote for reverting this change.

(We need to have another discussion outside of this PR whether the `--cluster-pool` arg needs to be removed entirely because I'm not sure this ever worked.)

File: flytekit/models/execution.py [hunk 3]
```diff
@@ -198,6 +200,7 @@ def __init__(
         :param overwrite_cache: Optional flag to overwrite the cache for this execution.
         :param envs: flytekit.models.common.Envs environment variables to set for this execution.
         :param tags: Optional list of tags to apply to the execution.
+        :param execution_cluster_label: Optional execution cluster to use for this execution.
```
Comment (fg91) [inline flytekit/models/execution.py:203]:
```suggestion
        :param execution_cluster_label: Optional execution cluster label to use for this execution.
```

File: flytekit/remote/remote.py [hunk 4]
```diff
@@ -1111,6 +1113,7 @@ def _execute(
         :param envs: Environment variables to set for the execution.
         :param tags: Tags to set for the execution.
         :param cluster_pool: Specify cluster pool on which newly created execution should be placed.
+        :param execution_cluster_label: Specify cluster on which newly created execution should be placed.
```
Comment (fg91) [inline flytekit/remote/remote.py:1116]:
```suggestion
        :param execution_cluster_label: Specify label of cluster(s) on which newly created execution should be placed.
```

File: flytekit/remote/remote.py [hunk 5]
```diff
@@ -1287,6 +1294,7 @@ def execute(
         :param envs: Environment variables to be set for the execution.
         :param tags: Tags to be set for the execution.
         :param cluster_pool: Specify cluster pool on which newly created execution should be placed.
+        :param execution_cluster_label: Specify cluster on which newly created execution should be placed.
```
Comment (fg91) [inline flytekit/remote/remote.py:1297]:
```suggestion
        :param execution_cluster_label: Specify label of cluster(s) on which newly created execution should be placed.
```

File: flytekit/remote/remote.py [hunk 6]
```diff
@@ -1686,6 +1713,7 @@ def execute_local_task(
         :param envs: Environment variables to set for the execution.
         :param tags: Tags to set for the execution.
         :param cluster_pool: Specify cluster pool on which newly created execution should be placed.
+        :param execution_cluster_label: Specify cluster on which newly created execution should be placed.
```
Comment (fg91) [inline flytekit/remote/remote.py:1716]:
```suggestion
        :param execution_cluster_label: Specify label of cluster(s) on which newly created execution should be placed.
```

File: flytekit/remote/remote.py [hunk 7]
```diff
@@ -1835,6 +1868,7 @@ def execute_local_launch_plan(
         :param envs: Environment variables to be passed into the execution.
         :param tags: Tags to be passed into the execution.
         :param cluster_pool: Specify cluster pool on which newly created execution should be placed.
+        :param execution_cluster_label: Specify cluster on which newly created execution should be placed.
```
Comment (fg91) [inline flytekit/remote/remote.py:1871]:
```suggestion
        :param execution_cluster_label: Specify label of cluster(s) on which newly created execution should be placed.
```

File: flytekit/remote/remote.py (hunk 4 above)
Comment (fg91) [reply flytekit/remote/remote.py:1116]: Nit:

From the [docs](https://docs.flyte.org/en/latest/deployment/deployment/multicluster.html):

```yaml
   labelClusterMap:
      label1:
      - id: dataplane_1
         weight: 1
      label2:
      - id: dataplane_2
         weight: 1
```

An execution cluster label can specify a load balancing config of 50% dataplane_1 and 50% dataplane_2.

I wonder whether we should reflect this in the wording here with e.g. `label of cluster(s)`.

**Review by fg91 — APPROVED** [review]
I was halfway through implementing the exact same thing because we need this as well when I figured I should check whether somebody else is working on this given that the backend recently added support for this.

So thank you for having added this functionality to flytekit 🙏 

I tested your PR in a GCP-based multi-cluster deployment and can confirm that the cluster selection works. (For the record: I didn't use `pyflyte` but `FlyteRemote` directly.)

I proposed a minor rewording suggestion for the doc strings, not blocking though, so LGTM.

**Review by fg91 — CHANGES_REQUESTED** [review]
Just saw that tests aren't passing yet, can you please fix this and request another review?

File: flytekit/clis/sdk_in_container/run.py (hunk 2 above)
Comment (va6996) [reply flytekit/clis/sdk_in_container/run.py:252]: I'm not sure why `--cluster-pool` wouldn't work? My reason for removing the latter was the missing `--`. I figured it is a ripple effect where the previous ones didn't have the `--` and so it was missed in the subsequent ones. Is it ok to have a config without `--`?

File: flytekit/remote/remote.py (hunk 4 above)
Comment (va6996) [reply flytekit/remote/remote.py:1116]: Fair point!

File: flytekit/clis/sdk_in_container/run.py (hunk 2 above)
Comment (fg91) [reply flytekit/clis/sdk_in_container/run.py]: > I'm not sure why --cluster-pool wouldn't work?

Not related to this PR, I was referring to [this](https://github.com/flyteorg/flyte/pull/4956#issuecomment-2114843886) discussion.

File: flytekit/clis/sdk_in_container/run.py [hunk 8]
```diff
@@ -256,6 +256,16 @@ class RunLevelParams(PyFlyteParams):
             help="Assign newly created execution to a given cluster pool",
         )
     )
+    execution_cluster_label: str = make_click_option_field(
+        click.Option(
+            param_decls=["--execution-cluster-label", "-ecl"],
```
Comment (fg91) [inline flytekit/clis/sdk_in_container/run.py:261]:
```suggestion
            param_decls=["--execution-cluster-label", "--ecl"],
```

Should there be two hyphens? See e.g. `["--envvars", "--env"]`.

Comment (va6996) [reply flytekit/clis/sdk_in_container/run.py:261]: My intent was to use one similar to how you'd have for `--version` = `-v`? Lmk what you think. `-ecl` [should work too](https://stackoverflow.com/a/75448996).

**Comment by davidmirror-ops** [comment]
@eapolinario any idea of how to fix the CI check?

File: flytekit/clis/sdk_in_container/run.py (hunk 8 above)
Comment (fg91) [reply flytekit/clis/sdk_in_container/run.py]: Looking at the existing args ...

```py
param_decls=["--envvars", "--env"],
param_decls=["--raw-output-data-prefix", "--raw-data-prefix"],
param_decls=["-r", "--remote"],
```

... I'd say currently we only use a single hyphen if it is followed by a single character like `-v` or `-r`. If multiple characters follow like `--env`, we use two hyphens.

I don't have strong opinions about this but I'd propose to just stick with `--` here as well.

**Comment by fg91** [comment]
> @davidmirror-ops , CI failure is being fixed in #2477

@va6996 the fix has been merged. Could you please rebase and also merge the suggestion in https://github.com/flyteorg/flytekit/pull/2422#discussion_r1638465659?

**Review by fg91 — APPROVED** [review]
Thank you, will be very useful 🙏

**Comment by davidmirror-ops** [comment]
@eapolinario @cosmicBboy @wild-endeavor @pingsutw if any of you can review, we can unblock this PR. Thanks!

**Review by eapolinario — APPROVED** [review]

**Comment by Future-Outlier** [comment]
Hi, @va6996 I can't run this PR on the remote flyte cluster.
Can you help me check if this is a bug on flytekit?

```python
from flytekit.loggers import logger
from flytekit import ImageSpec, Secret, task, workflow
from flytekit import Resources

@task(requests=Resources(cpu="800m", mem="900Mi"))
def say_hello() -> str:
    return "hello"

@workflow
def wf() -> str:
    return say_hello()

if __name__ == "__main__":
    print(f"Running say_hello() {say_hello()}")
    # rgister(wf, endpoint="localhost:30081", project="flyteworkshop", domain="development", name="say_hello")
```
image

**Comment by va6996** [comment]
> Hi, @va6996 I can't run this PR on the remote flyte cluster. Can you help me check if this is a bug on flytekit?

Could you please send the stack trace so I could debug? I wonder if flyteidl/flyteadmin needs to be updated to understand `execution_cluster_label`? Could you also run `pip show flyteidl` and check the version?

**Comment by Future-Outlier** [comment]
> Could you please send the stack trace so I could debug? I wonder if flyteidl/flyteadmin needs to b

## Flyte, flytekit and flyteidl version

flyte version: master branch with latest version
flytekit version: master branch with latest version
flyteidl version: master branch with latest version
```python
(dev) future@outlier ~ % pip list | grep -iE flyte   
flyteidl                                 1.11.0b2.dev39+g8fc782f25 /Users/future-outlier/code/flyte/flyteidl
flytekit                                 1.12.4.dev19+g876877abd
```

## Example
above

## Stack Trace Error Message
```python
(dev) future@outlier ~ % pyflyte --verbose run --remote task_example.py wf

Running Execution on Remote.
Verbose mode on
╭────────────────────────────────────────── Traceback (most recent call last) ──────────────────────────────────────────╮
│ /Users/future-outlier/miniconda3/envs/dev/bin/pyflyte:8 in <module>                                                   │
│                                                                                                                       │
│ ❱ 8 │   sys.exit(main())                                                                                              │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/click/core.py:1157 in __call__                  │
│                                                                                                                       │
│ ❱ 1157 │   │   return self.main(*args, **kwargs)                                                                      │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/rich_click/rich_command.py:126 in main          │
│                                                                                                                       │
│ ❱ 126 │   │   │   │   │   rv = self.invoke(ctx)                                                                       │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/clis/sdk_in_container/utils.py:148 in invoke                             │
│                                                                                                                       │
│ ❱ 148 │   │   │   │   raise e                                                                                         │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/clis/sdk_in_container/utils.py:142 in invoke                             │
│                                                                                                                       │
│ ❱ 142 │   │   │   return super().invoke(ctx)                                                                          │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/click/core.py:1688 in invoke                    │
│                                                                                                                       │
│ ❱ 1688 │   │   │   │   │   return _process_result(sub_ctx.command.invoke(sub_ctx))                                    │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/click/core.py:1688 in invoke                    │
│                                                                                                                       │
│ ❱ 1688 │   │   │   │   │   return _process_result(sub_ctx.command.invoke(sub_ctx))                                    │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/click/core.py:1688 in invoke                    │
│                                                                                                                       │
│ ❱ 1688 │   │   │   │   │   return _process_result(sub_ctx.command.invoke(sub_ctx))                                    │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/click/core.py:1434 in invoke                    │
│                                                                                                                       │
│ ❱ 1434 │   │   │   return ctx.invoke(self.callback, **ctx.params)                                                     │
│                                                                                                                       │
│ /Users/future-outlier/miniconda3/envs/dev/lib/python3.9/site-packages/click/core.py:783 in invoke                     │
│                                                                                                                       │
│ ❱  783 │   │   │   │   return __callback(*args, **kwargs)                                                             │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/clis/sdk_in_container/run.py:580 in _run                                 │
│                                                                                                                       │
│ ❱ 580 │   │   │   │   run_remote(                                                                                     │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/clis/sdk_in_container/run.py:448 in run_remote                           │
│                                                                                                                       │
│ ❱ 448 │   execution = remote.execute(                                                                                 │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/remote/remote.py:1337 in execute                                         │
│                                                                                                                       │
│ ❱ 1337 │   │   │   return self.execute_remote_wf(                                                                     │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/remote/remote.py:1518 in execute_remote_wf                               │
│                                                                                                                       │
│ ❱ 1518 │   │   return self.execute_remote_task_lp(                                                                    │
│                                                                                                                       │
│ /Users/future-outlier/code/flytekit/flytekit/remote/remote.py:1479 in execute_remote_task_lp                          │
│                                                                                                                       │
│ ❱ 1479 │   │   return self._execute(                                                                                  │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
TypeError: _execute() got an unexpected keyword argument 'execution_cluster_label'
```

**Comment by Future-Outlier** [comment]
> > Hi, @va6996 I can't run this PR on the remote flyte cluster. Can you help me check if this is a bug on flytekit?
> 
> Could you please send the stack trace so I could debug? I wonder if flyteidl/flyteadmin needs to be updated to understand `execution_cluster_label`? Could you also run `pip show flyteidl` and check the version?

Wait, maybe there's some error on my setup, Kevin can run it correctly
